torchgen/dest/dispatchless.py

```diff
@@ -0,0 +1,247 @@
+            ),
+        )
+        for g in grouped_native_functions
+        for f, _ in get_native_and_group_pair(g)
+        if len(f.composite[0]) > 0 and b.dispatch_key in f.composite[1]
+    }
+
+
+# convenience function for checking if a NativeFunction is registered to a
+# dispatch key.
+def has_registered_kernel(
+    index: BackendIndex, f: NativeFunction, g: Optional[NativeFunctionsGroup]
+) -> bool:
+    if g is not None and g.structured:
+        # if f is structured, then it will be registered for both Meta and
+        # CompositeExpicitAutograd (both not in STRUCTURED_DISPATCH_KEYS).
+        cond_meta = index.dispatch_key == DispatchKey.Meta
+
+        # structured kernel generation does not register out functions for
+        # CompositeExplicitAutograd.
+        cond_composite = (
+            index.dispatch_key == DispatchKey.CompositeExplicitAutograd
+            and not f.func.is_out_fn()
+        )
+        return index.has_kernel(g) or cond_meta or cond_composite
+    return index.has_kernel(f)
+
+
+@dataclass(frozen=True)
+class DispatchlessComposite:
+    # BackendIndex for the dispatch key we are generating code for.
+    backend_index: BackendIndex
+
+    # a graph of dependency relations between operators.
+    graph: CompositeGraph
+
+    @staticmethod
+    def new(
+        backend_index: BackendIndex,
+        grouped_native_functions: Sequence[Union[NativeFunction, NativeFunctionsGroup]],
+    ) -> "DispatchlessComposite":
+        graph = get_graph(grouped_native_functions, backend_index)
+        return DispatchlessComposite(backend_index, graph)
+
+    # figures out what namespace to dispatch.
+    def _dispatch_namespace(
+        self, f: NativeFunction, g: Optional[NativeFunctionsGroup]
+    ) -> Optional[DispatchKey]:
+        dispatch_key = self.backend_index.dispatch_key
+
+        # Bypass the dispatcher if:
+        #   1. There is a kernel registered for f in dispatch_key; and
+        #   2. The key we are currently generating code for is not
+        #      CompositeExplicitAutograd
+        if has_registered_kernel(self.backend_index, f, g) and (
+            dispatch_key != DispatchKey.CompositeExplicitAutograd
+        ):
+            return dispatch_key
+
+        # bail if we have found no registered kernel.
+        return None
+
+    def aggregated_headers(self) -> List[str]:
+        includes = set()
+        includes.add("#include <ATen/NativeFunctions.h>")
+
+        for _, info in self.graph.items():
+            for dep_f, dep_g in info.dependencies:
+                ns = self._dispatch_namespace(dep_f, dep_g)
+                prefix = f"{ns}" if ns is not None else ""
+                includes.add(f"#include <ATen/{prefix}Functions.h>")
+
+        return sorted(includes)
+
+    def operator_headers(self) -> List[str]:
+        includes = set()
+
+        for _, info in self.graph.items():
+            includes.add(f"#include <ATen/ops/{info.f.root_name}_native.h>")
```

Comment:
The latest error: mobile internal builds aren't happy that we're includeding `ops_native.h`, or `ATen/NativeFunctions.h`.

I think we should actually be fine getting rid of them though? The dispatchless code always calls one of `at::cpu::foo`, `at::cuda::foo`, `at::compositeexplicitautograd::foo`, or `at::foo` (if it has to fall back to the dispatcher). So we never actually need to directly invoke the `at::native::` kernel. Does that sound right?

Also heads up I made some minor tweaks to the PR, so you might need to pull first if you make changes (removed unnecessary headers, added a `pt_defs.oss.bzl` file. This is to appease internal CI.

Reply:
The `_native.h` header contains the declaration for the function being defined, so it is necessary. Note that it's `info.f` which is the composite function, not one of its dependencies.

Reply:
Ah yep, thanks @peterbell10!

Looks like internal builds currently just special-case these headers, and include them by doing `#include "NativeFunctions.h"` instead... I'm going to see if we can clean that up so it works more like OSS.